ekncontent/ekncontent/eknc-domain.c

```diff
@@ -739,36 +739,134 @@ static void
+   * allow the creator of the countdown to handle the error */
+  if (error && countdown->error_callback)
+    {
+      (*countdown->error_callback)(error, countdown->data);
+    }
+
+  if (--countdown->remaining == 0)
+    {
+      (*countdown->callback)(countdown->data);
+      g_free (countdown);
+      *countdown_ptr = NULL;
+    }
+}
+
+typedef struct {
+  /* Note that scatter_task_countdown_subtask_completed will free
+   * this member once the countdown has completed */
+  ScatterTaskCountdown *countdown;
+  EkncDomain *domain;
+  GPtrArray *models;
+  guint index;
+} ObjectResponseData;
+
+static ObjectResponseData *
+object_response_data_new (ScatterTaskCountdown *countdown,
+                          EkncDomain *domain,
+                          GPtrArray *models,
+                          guint index)
+{
+  ObjectResponseData *data = g_new0(ObjectResponseData, 1);
+  data->countdown = countdown;
+  data->domain = domain;
+  data->models = models;
+  data->index = index;
+  return data;
+}
+
+static void
+object_response_data_free (ObjectResponseData *data)
+{
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(ObjectResponseData, object_response_data_free);
+
 void
-on_object_response (GObject *source,
-                    GAsyncResult *result,
-                    gpointer user_data)
+on_received_all_model_objects (gpointer user_data)
 {
-  g_autoptr(GTask) task = user_data;
-  EkncDomain *domain = g_task_get_source_object (task);
-  QueryState *state = g_task_get_task_data (task);
-  GError *error = NULL;
+  GTask *task = user_data;
 
+  /* We might have already been marked as completed earlier due to an
+   * error state. Don't return if that is the case */
   if (g_task_get_completed (task))
     return;
 
-  EkncContentObjectModel *model = eknc_domain_get_object_finish (domain, result, &error);
+  g_task_return_boolean (task, TRUE);
+  g_object_unref (task);
+}
+
+void
+on_receive_model_error (GError *error, gpointer user_data)
+{
+  GTask *task = user_data;
+
   if (error != NULL)
     {
       g_task_return_error (task, error);
       return;
     }
+}
+
+void
+on_object_response (GObject *source,
+                    GAsyncResult *result,
+                    gpointer user_data)
+{
+  g_autoptr(ObjectResponseData) object_response_data = user_data;
+  EkncDomain *domain = object_response_data->domain;
+  GError *error = NULL;
 
-  state->models = g_slist_append (state->models, model);
+  EkncContentObjectModel *model = eknc_domain_get_object_finish (domain, result, &error);
+
+  *(&g_ptr_array_index (object_response_data->models, object_response_data->index)) = model;
+  scatter_task_countdown_subtask_completed (&object_response_data->countdown, error);
+}
 
-  if (g_slist_length (state->models) == state->total_models)
+void
+maybe_unref_object (gpointer data)
+{
+  if (data)
     {
-      g_task_return_boolean (task, TRUE);
-      return;
+      g_object_unref (G_OBJECT(data));
```

Comment:
This function should at the very least be `static`, but can't you avoid adding `NULL` to the array in the first place?
Also, you don't need to cast `data` to `G_OBJECT()` -- `g_object_unref()` takes a `void *`